File: pipelinereposerver/src/main/java/com/greenops/pipelinereposerver/repomanager/RepoManagerImpl.java
package com.greenops.pipelinereposerver.repomanager;

import com.greenops.pipelinereposerver.dbclient.DbKey;
import com.greenops.pipelinereposerver.kubernetesclient.KubernetesClient;
import com.greenops.util.datamodel.git.GitRepoSchema;
import com.greenops.util.dbclient.DbClient;
import lombok.extern.slf4j.Slf4j;
import org.apache.logging.log4j.util.Strings;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import static com.greenops.pipelinereposerver.repomanager.CommandBuilder.getFolderName;

@Slf4j
@Component
public class RepoManagerImpl implements RepoManager {

    private final String directory = "tmp";

    private final String orgName;
    private Set<GitRepoCache> gitRepos;

    @Autowired
    public RepoManagerImpl(DbClient dbClient, KubernetesClient kubernetesClient, @Value("${application.org-name}") String orgName) {
        this.gitRepos = new HashSet<>();
        this.orgName = orgName;
        if (!setupRepoCache(dbClient, kubernetesClient)) {
            throw new RuntimeException("The org's repos could not be cloned correctly. Please restart to try again.");
        }
        dbClient.shutdown();
    }

    @Override
    public String getOrgName() {
        return orgName;
    }

    @Override
    public Set<GitRepoCache> getGitRepos() {
        return gitRepos;
    }

    @Override
    public boolean clone(GitRepoSchema gitRepoSchema) {
        if (containsGitRepoSchema(gitRepoSchema)) {
            return true;
        }
        try {
            var command = new CommandBuilder()
                    .gitClone(gitRepoSchema)
                    .cd(getFolderName(gitRepoSchema.getGitRepo()))
                    .gitLog(1, true)
                    .build();
            var process = new ProcessBuilder()
                    .command("/bin/bash", "-c", command)
                    .directory(new File(orgName + "/" + directory))
                    .start();
            int exitCode = process.waitFor();
            if (exitCode == 0) {
                log.info("Cloning repo {} was successful.", gitRepoSchema.getGitRepo());
                var commitHash = new String(process.getInputStream().readAllBytes()).split("\n")[0];
                gitRepos.add(new GitRepoCache(commitHash, commitHash, gitRepoSchema));
                return true;
            } else {
                log.info("Cloning repo {} was not successful. Error: {}\nCleaning up...", gitRepoSchema.getGitRepo(), new String(process.getErrorStream().readAllBytes()));
                delete(gitRepoSchema);
                return false;
            }
        } catch (IOException | InterruptedException e) {
            log.info("An error was thrown when attempting to clone the repo {}", gitRepoSchema.getGitRepo(), e);
            delete(gitRepoSchema);
            return false;
        }
    }

    @Override
    public boolean update(GitRepoSchema gitRepoSchema) {
        //There should only be one answer that fits the filter for oldGitRepoSchema
        var oldGitRepoCache = gitRepos.stream().filter(gitRepoCache -> gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoSchema.getGitRepo())).findFirst().orElse(null);
        if (oldGitRepoCache == null) {
            return false;
        }

        gitRepos = gitRepos.stream().filter(gitRepoCache -> !gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoSchema.getGitRepo())).collect(Collectors.toSet());
        var newRepoCacheEntry = new GitRepoCache(oldGitRepoCache.getRootCommitHash(), getCurrentCommit(gitRepoSchema.getGitRepo()), gitRepoSchema);
        gitRepos.add(newRepoCacheEntry);
        if (sync(gitRepoSchema)) {
            return true;
        } else {
            gitRepos.remove(newRepoCacheEntry);
            gitRepos.add(oldGitRepoCache);
            return false;
        }
    }

    @Override
    public boolean delete(GitRepoSchema gitRepoSchema) {
        try {
            var command = new CommandBuilder()
                    .deleteFolder(gitRepoSchema)
                    .build();
            var process = new ProcessBuilder()
                    .command("/bin/bash", "-c", command)
                    .directory(new File(orgName + "/" + directory))
                    .start();
            int exitCode = process.waitFor();
            log.info("Deletion of repo {} finished with exit code {}", gitRepoSchema.getGitRepo(), exitCode);
            if (exitCode == 0) {
                gitRepos = gitRepos.stream().filter(gitRepoCache -> !gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoSchema.getGitRepo())).collect(Collectors.toSet());
                return true;
            } else {
                return false;
            }
        } catch (IOException | InterruptedException e) {
            log.error("An error was thrown when attempting to delete the repo {}", gitRepoSchema.getGitRepo(), e);
            return false;
        }
    }

    @Override
    public String getYamlFileContents(String gitRepoUrl, String filename) {
        var listOfSchemas = gitRepos.stream()
                .filter(gitRepoCache -> gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoUrl))
                .collect(Collectors.toList());
        if (listOfSchemas.size() != 1) {
            log.info("Too many git repos with the given url");
            return null;
        }

        try {
            var pathToRoot = strip(listOfSchemas.get(0).getGitRepoSchema().getPathToRoot(), '/');
            var truncatedFilePath = strip(filename, '/');
            var path = Paths.get(Strings.join(
                    List.of(orgName, directory, getFolderName(gitRepoUrl), pathToRoot, truncatedFilePath),
                    '/')
            );
            var reader = Files.newBufferedReader(path);
            var data = reader.lines().collect(Collectors.joining("\n"));
            reader.close();
            return data;
        } catch (IOException ex) {
            log.info("File was not found or could not be read.");
            return null;
        }
    }

    @Override
    public boolean sync(GitRepoSchema gitRepoSchema) {
        var listOfGitRepos = gitRepos.stream().filter(gitRepoCache ->
                gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoSchema.getGitRepo())).collect(Collectors.toList());
        if (listOfGitRepos.size() != 1) {
            //The size should never be greater than 1
            return false;
        }
        var cachedGitRepoSchema = listOfGitRepos.get(0).getGitRepoSchema();
        try {
            var command = new CommandBuilder()
                    .gitCheckout("main")
                    .gitPull(cachedGitRepoSchema)
                    .build();
            var process = new ProcessBuilder()
                    .command("/bin/bash", "-c", command)
                    .directory(new File(orgName + "/" + directory + "/" + getFolderName(cachedGitRepoSchema.getGitRepo())))
                    .start();
            int exitCode = process.waitFor();
            if (exitCode == 0) {
                log.info("Pulling repo {} was successful.", cachedGitRepoSchema.getGitRepo());
                var commitHash = getCurrentCommit(cachedGitRepoSchema.getGitRepo());
                if (commitHash == null) return false;
                gitRepos = gitRepos.stream().filter(gitRepoCache -> !gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoSchema.getGitRepo())).collect(Collectors.toSet());
                var updatedCacheSchema = listOfGitRepos.get(0);
                updatedCacheSchema.setCurrentCommitHash(commitHash);
                updatedCacheSchema.setRootCommitHash(commitHash);
                gitRepos.add(updatedCacheSchema);
                return true;
            } else {
                log.info("Pulling repo {} was not successful.", cachedGitRepoSchema.getGitRepo());
                return false;
            }
        } catch (IOException | InterruptedException e) {
            log.error("An error was thrown when attempting to sync the repo {}", cachedGitRepoSchema.getGitRepo(), e);
            delete(cachedGitRepoSchema);
            return false;
        }
    }

    @Override
    public boolean resetToVersion(String gitCommit, String gitRepoUrl) {
        var listOfGitRepos = gitRepos.stream().filter(gitRepoCache ->
                gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoUrl)).collect(Collectors.toList());
        if (listOfGitRepos.size() != 1) {
            //The size should never be greater than 1
            return false;
        }
        var cachedGitRepoSchema = listOfGitRepos.get(0).getGitRepoSchema();
        try {
            var command = new CommandBuilder()
                    .gitCheckout(gitCommit)
                    .build();
            var process = new ProcessBuilder()
                    .command("/bin/bash", "-c", command)
                    .directory(new File(orgName + "/" + directory + "/" + getFolderName(cachedGitRepoSchema.getGitRepo())))
                    .start();
            int exitCode = process.waitFor();
            if (exitCode == 0) {
                gitRepos = gitRepos.stream().filter(gitRepoCache -> !gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoUrl)).collect(Collectors.toSet());
                var updatedCacheSchema = listOfGitRepos.get(0);
                updatedCacheSchema.setCurrentCommitHash(gitCommit);
                gitRepos.add(updatedCacheSchema);
                log.info("Updating repo version {} was successful.", cachedGitRepoSchema.getGitRepo());
                return true;
            } else {
                log.info("Updating repo version {} was not successful.", cachedGitRepoSchema.getGitRepo());
                return false;
            }
        } catch (IOException | InterruptedException e) {
            log.error("An error was thrown when attempting to update the repo version {}", cachedGitRepoSchema.getGitRepo(), e);
            return false;
        }
    }

    @Override
    public String getLatestCommitFromCache(String gitRepoUrl) {
        var listOfGitRepos = gitRepos.stream().filter(gitRepoCache ->
                gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoUrl)).collect(Collectors.toList());
        if (listOfGitRepos.size() != 1) {
            //The size should never be greater than 1
            return null;
        }
        return listOfGitRepos.get(0).getCurrentCommitHash();
    }

    @Override
    public boolean containsGitRepoSchema(GitRepoSchema gitRepoSchema) {
        return gitRepos.stream().anyMatch(gitRepoCache -> gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoSchema.getGitRepo()));
    }

    @Override
    public String getRootCommit(String gitRepoUrl) {
        var listOfGitRepos = gitRepos.stream().filter(gitRepoCache ->
                gitRepoCache.getGitRepoSchema().getGitRepo().equals(gitRepoUrl)).collect(Collectors.toList());
        if (listOfGitRepos.size() != 1) {
            //The size should never be greater than 1
            return null;
        }
        return listOfGitRepos.get(0).getRootCommitHash();
    }

    @Override
    public String getCurrentCommit(String gitRepoUrl) {
        try {
            var command = new CommandBuilder()
                    .gitLog(1, true)
                    .build();
            var process = new ProcessBuilder()
                    .command("/bin/bash", "-c", command)
                    .directory(new File(orgName + "/" + directory + "/" + getFolderName(gitRepoUrl)))
                    .start();
            int exitCode = process.waitFor();
            if (exitCode == 0) {
                log.info("Fetching commit was successful.");
                return new String(process.getInputStream().readAllBytes()).split("\n")[0];
            } else {
                log.info("Fetching commit was not successful.");
                return null;
            }
        } catch (IOException | InterruptedException e) {
            log.error("An error was thrown when attempting to fetch the commit hash", e);
            return null;
        }
    }

    private String strip(String str, char delimiter) {
        var beg = 0;
        var end = str.length() - 1;
        while (beg < str.length() && str.charAt(beg) == delimiter) {
            beg++;
        }
        while (end >= 0 && str.charAt(end) == delimiter) {
            end--;
        }
        if (end < beg) {
            return "";
        } else {
            return str.substring(beg, end + 1);
        }
    }

    private boolean setupRepoCache(DbClient dbClient, KubernetesClient kubernetesClient) {
        var command = new CommandBuilder()
                .mkdir(orgName)
                .cd(orgName)
                .mkdir(directory)
                .cd(directory)
                .build();
        try {
            var process = new ProcessBuilder()
                    .command("/bin/bash", "-c", command)
                    .start();
            int exitCode = process.waitFor();
            if (exitCode != 0) {
                log.info("Errors: {}", new String(process.getErrorStream().readAllBytes()));
                return false;
            }
        } catch (IOException | InterruptedException e) {
            return false;
        }
        var listOfTeams = dbClient.fetchStringList(DbKey.makeDbListOfTeamsKey(orgName));
        if (listOfTeams == null) {
            log.info("No teams in org {}", orgName);
            return true;
        }
        log.info("Fetched all teams and cloning pipeline repos for org {}", orgName);
        for (var teamName : listOfTeams) {
            var teamSchema = dbClient.fetchTeamSchema(DbKey.makeDbTeamKey(orgName, teamName));
            if (teamSchema == null) {
                log.info("The team {} doesn't exist, so cloning will be skipped", teamName);
                continue;
            }
            for (var pipelineSchema : teamSchema.getPipelineSchemas()) {
                var gitRepoSchema = pipelineSchema.getGitRepoSchema();
                var secretName = DbKey.makeSecretName(teamSchema.getOrgName(), teamName, pipelineSchema.getPipelineName());
                var gitCred = kubernetesClient.fetchGitCred(secretName);
                if (gitCred == null) {
                    throw new RuntimeException("Could not recover gitCred from kubernetes secrets");
                }
                gitRepoSchema.setGitCred(gitCred);
                if (!clone(gitRepoSchema)) {
                    return false;
                }
            }
            log.info("Finished cloning pipeline repos for team {}", teamName);
        }
        return true;
    }
}

File: pipelinereposerver/repomanager/repo_manager_impl.go
package repomanager

import (
	"github.com/greenopsinc/pipelinereposerver/commandbuilder"
	"github.com/greenopsinc/util/db"
	"github.com/greenopsinc/util/git"
	"github.com/greenopsinc/util/kubernetesclient"
	"log"
	"os"
	"os/exec"
	"strings"
)

const (
	DefaultOrg string = "org"
	directory  string = "tmp"
)

type RepoManager interface {
	GetOrgName() string
	GetGitRepos() map[string]git.GitRepoSchema
	Clone(gitRepoSchema git.GitRepoSchema) bool
	Update(gitRepoSchema git.GitRepoSchema) bool
	Delete(gitRepoSchema git.GitRepoSchema) bool
	GetYamlFileContents(filename string, gitRepoSchema git.GitRepoSchema) string
	Sync(gitRepoSchema git.GitRepoSchema) string
	ResetToVersion(gitCommit string, gitRepoSchema git.GitRepoSchema) bool
	ContainsGitRepoSchema(gitRepoSchema git.GitRepoSchema) bool
	GetCurrentCommit(gitRepoURL string) string
}

type RepoManagerImpl struct {
	orgName  string
	gitRepos map[string]git.GitRepoSchema
}

func New(dbOperator db.DbOperator, kubernetesClient kubernetesclient.KubernetesClient, orgName string) RepoManager {
	gitReposMap := make(map[string]git.GitRepoSchema)
	repoManager := &RepoManagerImpl{
		orgName:  orgName,
		gitRepos: gitReposMap,
	}
	if !repoManager.setupRepoCache(dbOperator, kubernetesClient) {
		log.Fatal("The org's repos could not be cloned correctly. Please restart to try again.")
	}
	return repoManager
}

func (r *RepoManagerImpl) GetOrgName() string {
	return r.orgName
}

func (r *RepoManagerImpl) GetGitRepos() map[string]git.GitRepoSchema {
	return r.gitRepos
}

func (r *RepoManagerImpl) Clone(gitRepoSchema git.GitRepoSchema) bool {
	if r.ContainsGitRepoSchema(gitRepoSchema) {
		return true
	}
	command := commandbuilder.New().
		GitClone(gitRepoSchema).
		Cd(commandbuilder.GetFolderName(gitRepoSchema.GetGitRepo())).
		GitLog(1, true).
		Build()

	process := exec.Command("/bin/bash", "-c", command)
	process.Dir = r.orgName + "/" + directory
	output, err := process.Output()
	if err != nil {
		log.Printf("Cloning repo %s was not successful. Error: %s - %s\nCleaning up...", gitRepoSchema.GetGitRepo(), err, output)
		r.Delete(gitRepoSchema)
		return false
	} else {
		log.Printf("Cloning repo %s was successful.", gitRepoSchema.GetGitRepo())
		r.gitRepos[getGitRepoKey(gitRepoSchema)] = gitRepoSchema
		return true
	}
}

func (r *RepoManagerImpl) Update(gitRepoSchema git.GitRepoSchema) bool {
	oldGitRepo, ok := r.gitRepos[getGitRepoKey(gitRepoSchema)]
	if !ok {
		return false
	}
	delete(r.gitRepos, getGitRepoKey(gitRepoSchema))
	r.gitRepos[getGitRepoKey(gitRepoSchema)] = gitRepoSchema
	if r.Sync(gitRepoSchema) != "" {
		return true
	} else {
		delete(r.gitRepos, getGitRepoKey(gitRepoSchema))
		r.gitRepos[getGitRepoKey(oldGitRepo)] = oldGitRepo
		return false
	}
}

func (r *RepoManagerImpl) Delete(gitRepoSchema git.GitRepoSchema) bool {
	command := commandbuilder.New().
		DeleteFolder(gitRepoSchema).
		Build()

	process := exec.Command("/bin/bash", "-c", command)
	process.Dir = r.orgName + "/" + directory
	output, err := process.Output()
	if err != nil {
		log.Printf("An error was thrown when attempting to delete the repo %s - %s", err, output)
		return false
	} else {
		delete(r.gitRepos, getGitRepoKey(gitRepoSchema))
		return true
	}
}

func (r *RepoManagerImpl) GetYamlFileContents(filename string, gitRepoSchema git.GitRepoSchema) string {
	if _, ok := r.gitRepos[getGitRepoKey(gitRepoSchema)]; !ok {
		log.Printf("Repo does not exist in manager")
		return ""
	}
	truncatedFilePath := strings.Trim(filename, "/")
	pathToRoot := strings.Trim(gitRepoSchema.GetPathToRoot(), "/")
	var fullPath string
	if pathToRoot == "" {
		fullPath = strings.Join([]string{r.orgName, directory, commandbuilder.GetFolderName(gitRepoSchema.GetGitRepo()), truncatedFilePath}, "/")
	} else {
		fullPath = strings.Join([]string{r.orgName, directory, commandbuilder.GetFolderName(gitRepoSchema.GetGitRepo()), pathToRoot, truncatedFilePath}, "/")
	}
	dat, err := os.ReadFile(fullPath)
	if err != nil {
		log.Printf("Error reading file: %s", err)
		return ""
	}
	return string(dat)
}

func (r *RepoManagerImpl) Sync(gitRepoSchema git.GitRepoSchema) string {
	cachedGitRepoSchema, ok := r.gitRepos[getGitRepoKey(gitRepoSchema)]
	if !ok {
		log.Printf("Repo does not exist in manager")
		return ""
	}

	command := commandbuilder.New().GitCheckout("main").GitPull(cachedGitRepoSchema).Build()
	process := exec.Command("/bin/bash", "-c", command)
	process.Dir = r.orgName + "/" + directory + "/" + commandbuilder.GetFolderName(cachedGitRepoSchema.GetGitRepo())
	output, err := process.Output()
	if err != nil {
		log.Printf("Pulling repo %s was not successful: %s - %s", cachedGitRepoSchema.GetGitRepo(), err, output)
		return ""
	} else {
		log.Printf("Pulling repo %s was successful.", cachedGitRepoSchema.GetGitRepo())
		commitHash := r.GetCurrentCommit(cachedGitRepoSchema.GetGitRepo())
		if commitHash == "" {
			return ""
		}
		return commitHash
	}
}

func (r *RepoManagerImpl) ResetToVersion(gitCommit string, gitRepoSchema git.GitRepoSchema) bool {
	cachedGitRepoSchema, ok := r.gitRepos[getGitRepoKey(gitRepoSchema)]
	if !ok {
		log.Printf("Repo does not exist in manager")
		return false
	}

	command := commandbuilder.New().GitCheckout(gitCommit).Build()
	process := exec.Command("/bin/bash", "-c", command)
	process.Dir = r.orgName + "/" + directory + "/" + commandbuilder.GetFolderName(cachedGitRepoSchema.GetGitRepo())
	output, err := process.Output()
	if err != nil {
		log.Printf("Updating repo version %s was not successful: %s - %s", cachedGitRepoSchema.GetGitRepo(), err, output)
		return false
	} else {
		log.Printf("Updating repo version %s was successful", cachedGitRepoSchema.GetGitRepo())
		return true
	}
}

func (r *RepoManagerImpl) GetCurrentCommit(gitRepoUrl string) string {
	command := commandbuilder.New().GitLog(1, true).Build()
	process := exec.Command("/bin/bash", "-c", command)
	process.Dir = r.orgName + "/" + directory + "/" + commandbuilder.GetFolderName(gitRepoUrl)
	output, err := process.Output()
	if err != nil {
		log.Printf("Fetching commit was not successful: %s - %s", err, output)
		return ""
	} else {
		log.Printf("Fetching commit was successful")
		return strings.Split(string(output), "\n")[0]
	}
}

func (r *RepoManagerImpl) ContainsGitRepoSchema(gitRepoSchema git.GitRepoSchema) bool {
	if _, ok := r.gitRepos[getGitRepoKey(gitRepoSchema)]; ok {
		return true
	}
	return false
}

func getGitRepoKey(gitRepoSchema git.GitRepoSchema) string {
	return gitRepoSchema.GetGitRepo() + gitRepoSchema.GetPathToRoot()
}

func strip(str string, delimiter string) string {
	beg := 0
	end := len(str) - 1
	for beg < len(str) && str[beg:beg+1] == delimiter {
		beg++
	}
	for end >= 0 && str[end:end+1] == delimiter {
		end--
	}
	if end < beg {
		return ""
	} else {
		return str[beg : end+1]
	}
}

func (r *RepoManagerImpl) setupRepoCache(dbOperator db.DbOperator, kubernetesClient kubernetesclient.KubernetesClient) bool {
	defer dbOperator.Close()
	command := commandbuilder.New().
		Mkdir(r.orgName).
		Cd(r.orgName).
		Mkdir(directory).
		Cd(directory).Build()

	output, err := exec.Command("/bin/bash", "-c", command).Output()
	if err != nil {
		log.Printf("Errors: %s - %s", err, output)
		return false
	}

	dbClient := dbOperator.GetClient()
	defer dbClient.Close()
	listOfTeams := dbClient.FetchStringList(db.MakeDbListOfTeamsKey(r.orgName))
	if len(listOfTeams) == 0 {
		log.Printf("No teams in org %s", r.orgName)
		return true
	}
	log.Printf("Fetched all teams and cloning pipeline repos for org %s", r.orgName)
	for _, teamName := range listOfTeams {
		teamSchema := dbClient.FetchTeamSchema(db.MakeDbTeamKey(r.orgName, teamName))
		if teamSchema.GetOrgName() == "" {
			//means team is nil
			log.Printf("The team %s doesn't exist, so cloning will be skipped", teamName)
			continue
		}
		for _, pipelineSchema := range teamSchema.GetPipelineSchemas() {
			gitRepoSchema := pipelineSchema.GetGitRepoSchema()
			secretName := db.MakeSecretName(teamSchema.GetOrgName(), teamName, pipelineSchema.GetPipelineName())
			gitCred := kubernetesClient.FetchGitCred(secretName)
			if gitCred == nil {
				panic("Could not recover gitCred from kubernetes secrets")
			}
			gitRepoSchema.SetGitCred(gitCred)
			if !r.Clone(gitRepoSchema) {
				return false
			}
		}
		log.Printf("Finished cloning pipeline repos for team %s", teamName)
	}
	return true
}
